.circleci/config.yml

```diff
@@ -0,0 +1,193 @@
+          rm /etc/service/{postgresql,redis,template-renderer,website}/down && sv start postgresql redis template-renderer website
+          # Wait for the website to start.
+          sleep 5
+          sudo -u postgres createdb -O musicbrainz -T musicbrainz_test -U postgres musicbrainz_test_template
+          # As noted in docker/musicbrainz-tests/DBDefs.pm, CircleCI
+          # sets `NO_PROXY=127.0.0.1,localhost` in every container
+          # (and it's not clear how to change this), so the Selenium
+          # proxy doesn't work unless we make requests against a
+          # different hostname alias. We use `mbtest`, added above.
+          # This mostly only matters on the form action attributes
+          # under t/selenium/release-editor/seeds/, and a few URL
+          # checks; the rel="selenium.base links aren't used at all.
+          sudo -E -H -u musicbrainz find t/selenium -type f -name '*.html' -exec sed -i'' -e "s/localhost:5000/mbtest:5000/g" '{}' \;
+          export JUNIT_OUTPUT_FILE=junit_output/selenium.xml
+          sudo -E -H -u musicbrainz carton exec -- prove \
+              -I lib \
+              t/selenium.js \
+              --harness=TAP::Harness::JUnit \
+              -v
+      - store_test_results:
+          path: ./junit_output
+
+  build:
+    executor: musicbrainz-tests
+    steps:
+      - restore_cache:
+          keys:
+            - v1-source-{{ .Branch }}-{{ .Revision }}
+            - v1-source-{{ .Branch }}-
+            - v1-source-
+      - checkout
+      - run: |
+          git submodule sync
+          git submodule update --init
+          chown -R musicbrainz:musicbrainz .
+          sudo -E -H -u musicbrainz git gc
+      - save_cache:
+          key: v1-source-{{ .Branch }}-{{ .Revision }}
+          paths:
+            - ".git"
+
+      - run: |
+          cd postgresql-musicbrainz-collate && make && make install && cd -
+          cd postgresql-musicbrainz-unaccent && make && make install && cd -
+
+      - restore_cache:
+          key: v1-mmd-schema-{{ .Environment.MMD_SCHEMA_TAG }}
+      - run: |
+          cd /home/musicbrainz
+          if [[ ! -d mmd-schema ]]; then
+            git clone https://github.com/metabrainz/mmd-schema.git
+            cd mmd-schema
+          else
+            cd mmd-schema
+            git fetch origin
+          fi
+          git checkout $MMD_SCHEMA_TAG
+          cd ../musicbrainz-server
+      - save_cache:
+          key: v1-mmd-schema-{{ .Environment.MMD_SCHEMA_TAG }}
+          paths:
+            - /home/musicbrainz/mmd-schema
+
+      - restore_cache:
+          keys:
+            - v2-node-{{ checksum "yarn.lock" }}
+            - v2-node-
+      - run: sudo -E -H -u musicbrainz yarn
+      - save_cache:
+          key: v2-node-{{ checksum "yarn.lock" }}
+          paths:
+            - "node_modules"
+
+      - restore_cache:
+          keys:
+            - v3-carton-{{ checksum "cpanfile.snapshot" }}
+            - v3-carton-
```

Comment:
Another critical feature here to speed up our builds is caching the build data using [save_cache](https://circleci.com/docs/2.0/configuration-reference/#save_cache) and [restore_cache](https://circleci.com/docs/2.0/configuration-reference/#restore_cache). Here it uses the checksum of `cpanfile.snapshot` as the cache key, but also accepts any partial match on `v3-carton-` as a fallback. (`carton` would then fix any missing or outdated modules.)

Reply:
That we can push a new `musicbrainz-tests` docker image without breaking our build caches is very nice, too. But it's important to be aware that this can cause issues in rare cases where we update a library in the image while our build cache references the old version. The `v3-` prefix should be incremented whenever we update build libs in that way.